src/each/each.py

```diff
@@ -24,17 +27,164 @@
 class WorkInProgress:
     pid = attr.ib()
     started = attr.ib()
-    source_file = attr.ib()
+    work_item = attr.ib()
     out_file = attr.ib()
     err_file = attr.ib()
     status_file = attr.ib()
 
 
+class WorkItem(ABC):
+    """A thing to processed by ``Each``."""
+
+    """A name for this work item that can be used as a filename."""
+    name = NotImplemented
+
+    @abstractmethod
+    def exists(self):
+        """Whether or not this work item still exists."""
+
+    @abstractmethod
+    def as_input_file(self):
+        """This work item as a file descriptor.
+
+        This file descriptor is used as STDIN on a user-provided command.
+        """
+
+    @abstractmethod
+    def as_argument(self):
+        """This work item as a command-line argument."""
+
+    @abstractmethod
+    def write_in_file(self, path):
+        """Create a file at 'path' that contains the input data for this work item."""
+
+
+@attr.s()
+class FileWorkItem(WorkItem):
+    """A file to be processed by ``Each``."""
+
+    """A name for this work item that can be used as a filename."""
+    name = attr.ib()
+
+    """The location of the file on disk."""
+    path = attr.ib()
+
+    def exists(self):
+        """A file work item exists only if the file exists."""
+        return os.path.exists(self.path)
+
+    def as_input_file(self):
+        return os.open(self.path, os.O_RDONLY)
+
+    def as_argument(self):
+        return os.path.abspath(self.path)
+
+    def write_in_file(self, path):
+        """The ``in`` file is a symlink to the original file."""
+        os.symlink(os.path.abspath(self.path), path)
+
+
+@attr.s()
+class LineWorkItem(WorkItem):
+    """A line to be processed by ``Each``."""
+
+    """A name for this work item that can be used as a filename."""
+    name = attr.ib()
+
+    """The line itself, including line ending, as text."""
+    line = attr.ib()
+
+    def exists(self):
+        """Whether or not this work item still exists.
+
+        A line always exists.
+        """
+        return True
+
+    def as_input_file(self):
+        r, w = os.pipe()
+        os.write(w, self.line.encode("utf-8"))
```

Comment:
Doesn't this leave `w` dangling? Do we not need to close it at this point if we're to avoid a file descriptor leak?